### DISCUSSION TITLE
Buffs/Debuffs (Temporary Attributes / "Effective" Values)

### DISCUSSION THREAD

**Discussion by strikaco** [evennia/evennia#2406] [General Discussion]
# The Problem
Suppose "Mario" eats "mushroom." Let's say Mario's size should increase by x2...for 30 seconds.

### Scripts?
Using a script, we could:
* Read and store Mario's current size (`self.db.obj_size = self.obj.db.size`)
* Set `self.obj.db.size = self.obj.db.size * 2`
* ...and reverse the change when the script expires. 

We're making permanent changes to Mario's attributes though; if the script is killed and doesn't run the appropriate hook that reverts the changes, Mario will be stuck with elephantitis. We wouldn't even know for certain what his original size was supposed to be in order to reset him to his original state.

And suppose Mario eats another mushroom while the first `(base * 2)` is active. His already-doubled size...will double again `(base * 2 * 2)`. If he eats _another_, he'd be 8 times his starting size. To avoid this, we'd have to make sure only one script at a time could affect a particular attribute...which isn't necessarily desirable (if he eats a _poison_ mushroom, we'd want to modify the same attribute in order to _halve_ the _base_ value and have the _multipliers_ apply to his diminished stature).

### Stacks?
The best way I've come up with to deal with this is to create something like a "stack" that emulates a DB transaction-- somehow record a list of operations to apply against an attribute (+1, *0.5, *2, -5, etc.) sorted by priority and indexed with a reference object, then when the "effective" value of that attribute needs to be read, you replay the operations against the base value. When particular operations should no longer apply (because the applicant object expired), that modification is identified by reference to the applicant object and popped from the list. 

The problem I've run into with this is every mechanism in this engine wants to read directly from the AttributeHandler; there's no way to mutate values on-read (that I know of) unless you manually create properties on the typeclass or create a custom AttributeHandler (which is a nightmare given how overloaded the `get()` method is). 

Serializing this operation list (which references in-memory objects, not all of which are persistent or serializable) is also difficult. I had some success storing scripts/objects as keys in `ndb`, but it eventually chokes on itself when a script expires or the object is deleted since it renders existing keys unhashable and Python will refuse to deal with any of it at that point.

### Namespacing?
Another idea I had was to use attribute categories to namespace overrides stored on the object itself. 
* `Mario.db.size[None]` might be `1`. This is the "base" value.
* After eating the first mushroom, it starts script `#222`. 
* `Mario.db.size[mushroomscript#222]` is set to `2` by the script.
* Mario eats another mushroom, starting script `#223`.
* `Mario.db.size[mushroomscript#223]` is set to `2` by the script (it should have no knowledge of and should not attempt to build off of the value set by the other script).

This doesn't work so well for several reasons, the first of which being that the AttributeHandler is explicit about category and will continue to assert `size == 1` since it has no idea I expect these categorized values to _override_ the base-- or in what order (this solution does not make it clear to myself either-- we don't _necessarily_ want attribute decoration prioritized by ID/sequential order. If by narrative design Mario is "permanently cursed" to be tiny by Bowser, the curse should have highest priority-- immediately eating a handful of garbage after its application should not be a remedy).

And while I can reference modified attributes on Mario by the object that set it, this gets harder to juggle when "equipped" objects enter the picture that also modify the same attributes-- `#222` could have been applied by a Hat object that is equipped, _or_ a Script (I'd have to check _both_ tablespaces and match the key + id).

### Custom Handler?
Currently thinking about writing some sort of EffectiveAttributeHandler, but again-- unless I flawlessly extend AttributeHandler and replace it on the typeclass, nothing will read from a separate handler without rewriting all attribute-related locks and a ton of get/set commands.

### Extended Model?
Have also considered an "ExtendedAttribute" model linked to Attribute by a one-to-one field, that adds supporting fields (like `priority`, `ttl` and a GFK to the setter object so GenericRelation will conveniently cascade deletions of overwritten Attributes), but, again, nothing internal will read any of these fields and this requires a bunch of Django admin hackery to view/edit them inline.

I don't think settings currently allows one to specify an alternate Attribute model, should I simply extend Attribute itself. I'm guessing I'd have to fork the entire project to use a custom Attribute model.

### Other ideas?
I don't have any other feasible ideas, so I'm deferring to community wisdom :)

#### How are _you_ handling buffs/debuffs?

**Comment by Griatch** [comment]
For `develop` branch we are adding a new contrib of `Traits` heavily inspired by Whitenoise's Traits from Ainneve. This is essentially implementing a custom AttributeHandler as you allude to above - it uses a normal Attribute for storage under the hood. 

There are several types of Traits, but one capability is that it remembers its original value and you can add a 'modifier' to it to temporarily vary its value (usually between a min/max). A Trait can also have a 'rate' applied to have this value go up and down over time. One drawback I've thought of is that they don't work natively to set in in prototypes at this time. I should probably look into that. 

... So not something available in master yet, but you can read more about Traits here to see if you can get inspired: https://www.evennia.com/docs/1.0-dev/api/evennia.contrib.traits.html

**Reply by strikaco** [reply]
Have not tried it yet, but traits looks like it might come _really_ close to what I'm looking for. It appears to support the `current`/`modded` dichotomy.

It's unclear to me if this will integrate with the rest of the native systems though-- to access something like `hp` it looks like you have to drill through multiple levels of objects: `<obj>.traits.hp.current` or so? Locks won't check that; they'll look at `<obj>.hp` or `<obj>.obj.db.hp` at deepest. (It'd work if the TraitHandler always stores the current value on `<obj>.db.hp` though...)

Will have to give it a try; thanks for the suggestion!

**Comment by Griatch** [comment]
Yes, one would need a custom lockfunc to check a trait. The trait data is stored internally as a dict on the Attribute. It's not much of a limit, considering how easy it is to add new lockfuncs. Maybe the contrib should provide some custom lockfuncs to easily add.

**Comment by strikaco** [comment]
While poking around at #1954, I thought of another solution.

### Shared Attribute Stack

Attributes are currently related to Objects through a many-to-many relationship instead of many-to-one. This struck me as an odd way of defining relationships for an object type intended to be unique per-parent, but it does imply a single Attribute instance can be shared by multiple parents.

So if _Attributes_ can be shared...and Scripts are _temporary_ objects...

```
class Decoration(Script):
    """
    Temporarily adds attributes from self to the bound object's stack.

    """
    def activate(self):
        """
        Share all attributes with bound object.
        
        """
        self.obj.db_attributes.add(*self.db_attributes.all())
    
    def deactivate(self):
        """
        Remove all shared attributes from bound object.
        
        """
        self.obj.db_attributes.remove(*self.db_attributes.all())
    
    def at_start(self, *args, **kwargs):
        self.activate(*args, **kwargs)
        return super().at_start(*args, **kwargs)
        
    def at_stop(self, *args, **kwargs):
        self.deactivate(*args, **kwargs)
        return super().at_stop(*args, **kwargs)
        
    def delete(self, *args, **kwargs):
        self.deactivate(*args, **kwargs)
        return super().delete(*args, **kwargs)
```

As a proof-of-concept it lacks any sort of fine-grained per-key configurability, but all Attributes defined on this very-simple Script will be non-destructively added to the bound Object's Attribute queryset/stack while the Script is active.

When the Script is stopped/deleted, those same Attributes are quietly removed from the bound Object's Attribute queryset/stack (since they were uniquely-defined on this Script, there is no chance of cascading deletion weirdness).

* `attredit` locks are of importance here, otherwise the bound Object can itself overwrite its own decorated values.
* This does result in duplicate-keyed Attributes being stored on the bound Object (I doubt the AttributeHandler knows what to do here, so you'd have to manually manage Attributes using `obj.db_attributes.all()`).
* There is no way to prioritize Attributes except by `id` (can't define custom fields-- hence #2418).
* It is not clear who the original owner of any given Attribute is supposed to be without a complex query (can't set metadata-- again, #2418).
* This defies the `db_model` field on the Attribute so I don't know what the consequences of commingling Attributes across two different object types might be.

**Reply by strikaco** [reply]
On further thought, shared attributes might be a bit of a risky proposition:

https://github.com/evennia/evennia/blob/56b1c259f742134a12e1440cae9fe522b75aefe9/evennia/typeclasses/models.py#L73-L77

This seems to imply that if you have two objects sharing an Attribute, deletion of either object will also delete the shared Attribute from both.

**Reply by Griatch** [reply]
Attributes use M2Mkeys since they are used for linking to different entities under the hood (Objects, Accounts). This makes it easier to add Attributes to any new user-made model as long as it has the right parent model. 

I don't think it's a good idea to share Attributes between objects. I'd rather see an extension of Scripts (for example) for expressing arbitrary relations between entities. But overall, Traits, with some extensions to make them fit better into the ecosystem, would probably fit better for the things you started this thread discussing.
